src/model/http/har.ts

```diff
@@ -141,44 +138,53 @@ export function generateHarRequest(
 
 type TextBody = {
     mimeType: string,
-    text: string
+    encoding?: string,
```

Comment:
Unfortunately, the dilemma here is whether to keep supporting HAR 1.2 spec or not. Personally, in my use-case I'm fine with implementing custom logic in my project to account for optional `_` prefixed fields containing base64 encoded request payloads, but, I have a feeling that this, including this semi-compliant HAR 1.3 PR, would be a lazy work-around of an issue that could potentially be brought up again in the future, by users who need the payload data but don't have means of implementing custom logic accounting for optional fields, into their work-environment/scenarios

As for cross-compatibility with HAR 1.2 and major HAR-based tools, I don't think there's any viable ways to make any upcoming changes work with current HAR 1.2 implementations on Firefox (and probably Chrome, haven't checked how it handles HAR files, but I'm fairly certain it's abiding to HAR 1.2)

Here's Firefox's variant of [postData builder](https://searchfox.org/mozilla-central/rev/253ae246f642fe9619597f44de3b087f94e45a2d/devtools/client/netmonitor/src/har/har-builder.js#270)

This exact issue with the 1.2 spec has been [brought up in the past](https://code.google.com/archive/p/http-archive-specification/issues/6), however, judging from the lack of activity, dating back to the google code thread, and even on the HAR 1.3 spec proposal repository, I'd say har is abandoned, and any attempts to maintain it are useless (previous attempts have failed).

Another idea would be to implement https://github.com/httptoolkit/httptoolkit/issues/104 and stop trying to work around 1.2 spec limitations, but that's a different discussion in itself.

Open to further input on this.

Reply:
All good points, and I totally agree that HAR has big problems and is sadly not maintained :disappointed: 

It's not so much about the theoretical principle of following the specification though (that would be nice, but it's not important). My larger concern is about practical compatibility with the other tools that import HARs. HARs is very widely used, and breaking compatibility with other tools for this will affect a lot of people's workflows. Following the spec where possible is a good way to guarantee (more or less) that compatibility.

I'd prefer to add custom fields because of that issue, since custom fields shouldn't affect usage with any existing tools, and will just be ignored if they're not understood.

Changes like this though will almost definitely break importing HARs from HTTP Toolkit into other tools: any HAR exports including the base64 data from this PR will either be rejected completely, if tools validate against schema for invalid fields, or they'll be imported but with the base64 bodies completely misinterpreted (because there's no way for other tools to know that this data is base64 - the current code is changing the semantics of `postData.text` in an incompatible way).

We can't ship something that will break compatibility with existing tools.

> Another idea would be to implement httptoolkit/httptoolkit#104

Very difficult to do unfortunately. PCAP is a much lower-level format: it's not storing HTTP requests, it's storing raw network traffic. That brings lots of problems, for example HTTPS traffic would need to be stored in the original encrypted form, including the raw client & server handshake, with the decryption keys used for the TLS session provided separately to allow decryption of the raw packets, etc etc. Most of the relevant data to do all this isn't preserved by HTTP Toolkit internally. PCAP also isn't very easy to handle if you're just interested in the HTTP content at all, JSON is much much more convenient for quick hacking. PCAP is great for low-level analysis of individual packets, but not so much for HTTP.

> I have a feeling that this, including this semi-compliant HAR 1.3 PR, would be a lazy work-around of an issue that could potentially be brought up again in the future, by users who need the payload data but don't have means of implementing custom logic accounting for optional fields

The `postBody.encoding` approach won't work for them either though, right? They'll import these HAR files and the bodies won't be decoded correctly, because they won't understand that field, so they're still stuck without custom logic.

For users who can write custom logic, we can basically do whatever we like, that part is nice and easy.

For users who can't use custom logic though (anybody with an existing custom workflow built on HAR files, or anybody importing HARs into other popular tools) we need something that stays compatible with HAR 1.2.

I'd like to aim for a progressive enhancement model: stick to the standards everywhere, do the best we can for tools that only support those, and then add extra custom fields in some sensible format (very happy to debate the specifics of that) to make it easy for custom logic to go further and get the full details.

We do already do this in some places, e.g. we add a `_tlsErrors` custom field that includes all TLS errors so those are available for custom analysis but they aren't included as normal-but-broken requests that might confuse other tools.

Reply:
>The postBody.encoding approach won't work for them either though, right? They'll import these HAR files and the bodies won't be decoded correctly, because they won't understand that field, so they're still stuck without custom logic.

Unfortunately, yes. Unless they happen to make use of some tool that implements HAR 1.3 (very unlikely)

>I'd prefer to add custom fields because of that issue, since custom fields shouldn't affect usage with any existing tools, and will just be ignored if they're not understood.

Custom fields sound like the best alternative in this case, all options considered.

>I'd like to aim for a progressive enhancement model: stick to the standards everywhere, do the best we can for tools that only support those, and then add extra custom fields in some sensible format (very happy to debate the specifics of that) to make it easy for custom logic to go further and get the full details.

```js
"_requestContent": {
    "data": "", // Data encoded according to the "encoding" field
    "size": "", // Needed since "entry.request.bodySize" would be -1 after UTF-8 decode failure, (although, not strictly enforced by 1.2 spec)
    "encoding": "base64", // Self-explanatory
    "comment": "" // Could probably be useful in the future
}
```
Semantics aside, seems fine for a start. Open for further suggestions.

Reply:
> Semantics aside, seems fine for a start. Open for further suggestions.

Looks like a good start to me! A couple of minor points:

> `_requestContent`

Can we call this just `_content`? It's already nested in the request, so that's a bit redundant.

> `data`

I know it's slightly weird, but I think we should probably name `data` as `text` instead, because then it's compatible with the [`response.content`](http://www.softwareishard.com/blog/har-12-spec/#content) property, which is convenient for parsing this later.

> "entry.request.bodySize" would be -1 after UTF-8 decode failure

Not quite - it should be the encoded size, i.e. `request.body.encoded.byteLength` in HTTP Toolkit. That's either the raw byte size of the body (in most cases) or smaller (if gzip or other compression was used for the request body). This is the size of the body on the wire, and I think we should still set it correctly even if the body can't be represented in UTF-8.

That said, you're right and we do still want this `size` field here too, to provide the _decoded_ size - after any decompression that might've been used, i.e. `request.body.decoded.byteLength` in HTTP Toolkit.

Sound reasonable? If you're happy to add that here then that would be great!

It'd be useful to do the equivalent to allow importing this data too in `parseHarRequest` if you have a sec, but if not I'm happy to merge it without and do that myself later, since the current export for these cases would be broken today anyway.

Reply:
>Not quite - it should be the encoded size, i.e. request.body.encoded.byteLength in HTTP Toolkit. That's either the raw byte size of the body (in most cases) or smaller (if gzip or other compression was used for the request body). This is the size of the body on the wire, and I think we should still set it correctly even if the body can't be represented in UTF-8

Good point, I was under the impression that the spec suggested defaulting to -1 in case of failure, but it turns out `bodySize` is defined as `Size of the request body (POST data payload) in bytes. Set to -1 if the info is not available.`

>It'd be useful to do the equivalent to allow importing this data too in parseHarRequest if you have a sec, but if not I'm happy to merge it without and do that myself later, since the current export for these cases would be broken today anyway.

I'd be more than happy to help you with this. Will adapt har import/export functionality accordingly on upcoming patch :+1:

Reply:
Addressed in 4c7f0f9f44074db80ae19f572b88446f50c3c39e